### ISSUE TITLE
Mail Sender Service

### ISSUE THREAD

**Issue by dchopra001** [dchopra001/Dym_hrk#4]
We need to develop a web service that will send different types of emails to clients based on daily events. This ties closely with how the DB is structured and where we are hosting the server. There should be a general layout for this which we can leverage in the future for all other projects. 

This is gonna end up being a pretty large item that we'll probably re-tool many times. I'll code up something to get us started this weekend.

**Comment by dchopra001** [comment]
Okay so for the purposes of this project, I have most of the design decided for the web service. I'm almost done looking into the technical details to figure out how to do everything I want. IMO it's best to implement this as a component of the entire application as opposed to a separate thing. This way the service only goes down if the entire site goes down. There's good and bad in that decision.. but given some of the restrictions on the hosting site, I think this is the way to go for now. 

With regards to a general overview of the service:
-Every 12 hours the service will check if another email needs to be sent out. This guarantees that we never miss a Birthday. Perhaps I can make this so that we always send the email out during the day time. 

-The service where she would like to send email notifications should be easy to write. There might be some issues with text formatting. But that we will have to test and iterate over. 

@liy4502 FYI. Let me know if you have any thoughts regarding this.

**Comment by liy4502** [comment]
We should divide this into 2 separate issues because one relates to the database, the other one is a service built into the application itself. 

For the request mail service:
I would just use the microsoft logging and use the smtp from the host to send the log (the request). 

Are you going to check the DB every 12 hour based on the system time? I would like a bit more detail regarding how you are checking.

**Comment by dchopra001** [comment]
> Are you going to check the DB every 12 hour based on the system time? I would like a bit more detail regarding how you are checking.

Yes. It will use the time provided by the system + the date to make sure we don't do the work twice. If the system time is not to be trusted then we'll need to grab it from an external time service. 

Considering how big the data will become in a few years, we need to do this in a stored procedure. For the stored procedure the input is the current date, and the output is a table of parent name + kid name + parent email + standard email message. There is going to be some admin pages required for this: Registration form for family (filled in at time of photo session), email draft screen where she can write a new email--> not sure about specifying attachments yet. This part can be tricky because we need to setup a preview for her that can **accurately** represent what the email will look like.

**Comment by dchopra001** [comment]
I forgot to mention there should be a method for subscribers to be able to unsubscribe from emails. Should be easy to implement: Web Request (or maybe a service?) that deletes a particular email address from a list.

**Comment by liy4502** [comment]
I think this is important. Regarding this, is your plan to incorporate this on release or backlog it due to the frequency of mail (3month, 6month, 1 year, therefore not really spammy)?

**Comment by liy4502** [comment]
Also are each user(parent child) going to register or is she going to put this information in herself?

**Comment by dchopra001** [comment]
>I think this is important. Regarding this, is your plan to incorporate this on release or backlog it due to the frequency of mail (3month, 6month, 1 year, therefore not really spammy)?

Are you referring to the ability to unsubscribe? This should be done right away. It's easy anyway. 

>Also are each user(parent child) going to register or is she going to put this information in herself?

That's a good question. I had thought that it should be her doing this via her Admin Access. A few reasons why:
1. She will be forced to regularly log into the site, hence she will realize it's an effective part of her business.
2. The more the App gets used, the more it's stressed, the sooner bugs are exposed (and fixed!)
3. I can't imagine people will register themselves if she tells them to. After people have their pictures, telling them to go register on the site is the equivalent on me going to buy glasses at a store and then the store clerk telling me to go on a site and give a good review to them via a survey... It'll never happen.

But it may be a good idea to discuss something like this over the phone with her. Ultimately she knows her business best.

**Comment by dchopra001** [comment]
To be more clear, my vision is similar to when you go to a big brand clothes store. When you're cashing out, they always ask for your email address and phone number. Then they scan and get you to pay. 

She will have a similar workflow. When people are paying, she can first ask them for the standard info for emails.
